### PR TITLE
additional SBM features to nurbs_curve_on_surface_geometry

### PR DESCRIPTION
**Pull request by NickNick9**
**📝 Description**
We have added to the nurbs_curve_on_surface_geometry the possibility of considering an SBM brep surface.
Instead of using the method ComputeAxisIntersection, in the SBM case we use a different one called ComputeAxisIntersectionSBM (that only looks for intersections on the knot vertices).

Moreover, we have also added the SBM case in the CreareQuadraturePoints method in the same file.

IgaApplication
Geometry

### CODE REVIEW

File: kratos/geometries/nurbs_curve_on_surface_geometry.h [hunk 1]
```diff
@@ -260,11 +260,127 @@ class NurbsCurveOnSurfaceGeometry : public Geometry<typename TSurfaceContainerPo
...
+        if ((physical_coord_2[0] - physical_coord_1[0] < 0 ) || 
+            (physical_coord_2[1] - physical_coord_1[1] < 0 )) {
+            reverse_order = true;
+        }
+        std::vector<double> tempSpans;
+        const double tolerance_orientation = 1e-10;
+        const double tolerance_intersection = 1e-15;
+        
+        // Scale factor between volume coordinate and surface one
+        double physical_length_segment = norm_2(physical_coord_1-physical_coord_2);
```
Comment (AlejandroCornejo) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h:359]: const?

File: kratos/sources/variables.cpp [hunk 2]
```diff
@@ -477,6 +477,9 @@ KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(PARAMETER_2D_COORDINATES)
 KRATOS_CREATE_VARIABLE(double, VARIATIONAL_REDISTANCE_COEFFICIENT_FIRST)
 KRATOS_CREATE_VARIABLE(double, VARIATIONAL_REDISTANCE_COEFFICIENT_SECOND)
 
+// SBM variables
+KRATOS_CREATE_VARIABLE(bool, IS_SBM)
```
Comment (AlejandroCornejo) [inline kratos/sources/variables.cpp]: maybe `IS_SHIFTED_BOUNDARY` ?

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I need to do a further review, but at first glance, I'm not a big fan of adding a geometry that depends on the method... In other words, I don't really like the `IS_SBM` stuff. Can you please justify its need?

**Comment by NickNick9** [comment]
@rubenzorrilla  > I need to do a further review, but at first glance, I'm not a big fan of adding a geometry that depends on the method... In other words, I don't really like the `IS_SBM` stuff. Can you please justify its need?

@NickNick9 
The fact is that our SBM stuff is very similar to the IBRA one. So we basically use the same geometries, but at the very end, we only cut along the knot vertices. So in more than one place, we have to differentiate between the two (Create an additional function SpansLocalSpace_SBM would be complex because it is called in many other places and still we would need a flag I guess). We are planning to use the IS_SBM when we create the gauss points for example.

So my reason is: SBM and IBRA conceptually are very similar, we want a flag that says which approach you want to use when you have to deal with the un-fitted boundary in IGA.

File: kratos/geometries/nurbs_curve_on_surface_geometry.h [hunk 3]
```diff
@@ -249,6 +249,17 @@ class NurbsCurveOnSurfaceGeometry : public Geometry<typename TSurfaceContainerPo
         this->SpansLocalSpace(rSpans, interval.GetT0(), interval.GetT1());
     }
 
+    /* @brief Provides intersections of the nurbs curve with the knots of the surface,
+     *         using the interval of this curve in the SBM case.
+     * @param vector of span intervals.
+     * @param index of chosen direction, for curves always 0.
+     */
+    void SpansLocalSpaceSBM(std::vector<double>& rSpans, IndexType DirectionIndex = 0) const
```
Comment (rubenzorrilla) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]:
```suggestion
    void SpansLocalSpaceSBM(
        std::vector<double>& rSpans,
        IndexType DirectionIndex = 0) const
```
Please follow the [style guide](https://kratosmultiphysics.github.io/Kratos/pages/Kratos/For_Users/Conventions/Style_Guide.html).

File: kratos/geometries/nurbs_curve_on_surface_geometry.h [hunk 4]
```diff
@@ -260,24 +271,30 @@ class NurbsCurveOnSurfaceGeometry : public Geometry<typename TSurfaceContainerPo
...
-                rSpans,
-                *(mpNurbsCurve.get()), Start, End,
-                surface_spans_u, surface_spans_v,
-                1e-6);
-        }
+    /* @brief  Provides intersections of the nurbs curve with the knots of the surface in the SBM case.
+     * @return vector of interval limitations.
+     */
+    void SpansLocalSpaceSBM(std::vector<double>& rSpans,
+        double Start, double End) const
```
Comment (rubenzorrilla) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]:
```suggestion
    void SpansLocalSpaceSBM(
        std::vector<double>& rSpans,
        double Start,
        double End) const
```
Same comment.

File: kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp [hunk 5]
```diff
@@ -772,9 +766,8 @@ typedef Node NodeType;
         brep_coordinates[1][0] = 2.0;
         brep_coordinates[1][1] = 0.0;
         
-        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
+        NurbsCurveOnSurfaceGeometry<3, PointerVector<Point>, PointerVector<Point>> curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp]: Can't you continue using the `auto` type?
```suggestion
        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```

File: kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp [hunk 6]
```diff
@@ -708,17 +703,16 @@ typedef Node NodeType;
         brep_coordinates[1][0] = 2.0;
         brep_coordinates[1][1] = 1.0;
         
-        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
+        NurbsCurveOnSurfaceGeometry<3, PointerVector<Point>, PointerVector<Point>> curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp]: Can't you continue using the `auto` type?
```suggestion
        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```

File: kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp [hunk 7]
```diff
@@ -600,15 +600,13 @@ typedef Node NodeType;
         brep_coordinates[1][0] = 0.0;
         brep_coordinates[1][1] = 3.0;
         
-        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
+        NurbsCurveOnSurfaceGeometry<3, PointerVector<Point>, PointerVector<Point>> curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp]: Can't you continue using the `auto` type?

```suggestion
        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```

File: kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp [hunk 8]
```diff
@@ -638,15 +636,13 @@ typedef Node NodeType;
         brep_coordinates[1][0] = 2.0;
         brep_coordinates[1][1] = 0.0;
         
-        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
+        NurbsCurveOnSurfaceGeometry<3, PointerVector<Point>, PointerVector<Point>> curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp]: Can't you continue using the `auto` type?

```suggestion
        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```

File: kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp [hunk 9]
```diff
@@ -674,13 +670,12 @@ typedef Node NodeType;
         brep_coordinates[1][0] = 2.0;
         brep_coordinates[1][1] = 1.0;
         
-        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
+        NurbsCurveOnSurfaceGeometry<3, PointerVector<Point>, PointerVector<Point>> curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/geometries/test_nurbs_curve_on_surface_geometry.cpp]: Can't you continue using the `auto` type?

```suggestion
        auto curve_on_surface = GenerateReferenceNurbs2dforKnotIntersections(brep_coordinates);
```

File: kratos/sources/variables.cpp [hunk 10]
```diff
@@ -957,8 +954,6 @@ void KratosApplication::RegisterVariables() {
     // Variational redistance
     KRATOS_REGISTER_VARIABLE(VARIATIONAL_REDISTANCE_COEFFICIENT_FIRST)
     KRATOS_REGISTER_VARIABLE(VARIATIONAL_REDISTANCE_COEFFICIENT_SECOND)
-
-    // SBM variables
-    KRATOS_REGISTER_VARIABLE(IS_SBM)
+
```
Comment (rubenzorrilla) [inline kratos/sources/variables.cpp]: Do `checkout master variables.cpp` and commit to revert the file to master status.

**Review by rubenzorrilla — COMMENTED** [review]
Overall makes sense. Nevertheless, I'd indicate the upcoming changes with code comments. Specially the unification of the SBM and the untrimmed cases, which is something that IMO must be done for sure (most probably in a future PR).

File: kratos/includes/variables.h [hunk 11]
```diff
@@ -485,6 +485,8 @@ namespace Kratos
     KRATOS_DEFINE_VARIABLE(double, VARIATIONAL_REDISTANCE_COEFFICIENT_FIRST)
     KRATOS_DEFINE_VARIABLE(double, VARIATIONAL_REDISTANCE_COEFFICIENT_SECOND)
 
+    // SBM variables
+    KRATOS_DEFINE_VARIABLE(bool, IS_SBM)
```
Comment (rubenzorrilla) [inline kratos/includes/variables.h]: These are no longer needed right?

File: kratos/sources/variables.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/sources/variables.cpp]: I think is no longer needed.

File: kratos/sources/variables.cpp [hunk 12]
```diff
@@ -954,6 +957,8 @@ void KratosApplication::RegisterVariables() {
     // Variational redistance
     KRATOS_REGISTER_VARIABLE(VARIATIONAL_REDISTANCE_COEFFICIENT_FIRST)
     KRATOS_REGISTER_VARIABLE(VARIATIONAL_REDISTANCE_COEFFICIENT_SECOND)
-    
+
+    // SBM variables
+    KRATOS_REGISTER_VARIABLE(IS_SBM)
```
Comment (rubenzorrilla) [inline kratos/sources/variables.cpp]: I think is no longer needed.

Comment (andrewgorgi) [reply kratos/sources/variables.cpp]: Is it possible that you are referring to an old commit? The last commit should be without any modification on the variables.h/.cpp files

Comment (rubenzorrilla) [reply kratos/sources/variables.cpp]: You're right. My notifications got crazy.

**Comment by rubenzorrilla** [comment]
> Overall makes sense. Nevertheless, I'd indicate the upcoming changes with code comments. Specially the unification of the SBM and the untrimmed cases, which is something that IMO must be done for sure (most probably in a future PR).

Did you already add the comments I mentioned in here? (I did not find them).

**Review by rubenzorrilla — APPROVED** [review]
All my initial concerns have been addressed (aside of the unification of the traditional and SBM integration points creation, which will be done in an upcoming PR).

**Pull request merged**